sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Canonicalize.scala

```diff
@@ -85,6 +87,10 @@ object Canonicalize {
     case Not(GreaterThanOrEqual(l, r)) => LessThan(l, r)
     case Not(LessThanOrEqual(l, r)) => GreaterThan(l, r)
 
+    // order the list in the In operator
+    case In(value, list) =>
+      In(value, list.sortBy(_.semanticHash()))
```

Comment:
Why using `semanticHash` instead of `hashCode`? Do you have a test case to show the difference?

Reply:
The only difference is that the elements in the list are canonicalized before the hash.  I can't think of any meaningful example. The only one I can think of is something like ` mybool in (5 < 2)` and ` mybool in (not 5 >= 2)`. But in the future we may have more rules here making meaningful using the `semanticHash` which is logically what we want here IMHO

Reply:
A few general suggestions. 
- We need to add test cases when we want to cover the extra cases. 
- Update the comments in the description when we add new code
- Try our best to make the code simple.
- Find the best test suite when we add a new test. If unable to find a good one, create a new test suite.

Reply:
thanks for your great comment. This is very helpful and I'll keep in mind for the future. I think I addressed all the points now. Thanks.